ostd/src/mm/page/allocator.rs

```diff
@@ -5,70 +5,86 @@
 //! TODO: Decouple it with the frame allocator in [`crate::mm::frame::options`] by
 //! allocating pages rather untyped memory from this module.
 
-use alloc::vec::Vec;
+use alloc::boxed::Box;
+use core::alloc::Layout;
 
 use align_ext::AlignExt;
-use buddy_system_allocator::FrameAllocator;
-use log::info;
+use log::{info, warn};
 use spin::Once;
 
-use super::{cont_pages::ContPages, meta::PageMeta, Page};
 use crate::{
-    boot::memory_region::MemoryRegionType,
-    mm::{Paddr, PAGE_SIZE},
+    mm::{
+        page::{meta::PageMeta, ContPages, Page},
+        Paddr, PAGE_SIZE,
+    },
     sync::SpinLock,
 };
 
-/// FrameAllocator with a counter for allocated memory
-pub(in crate::mm) struct CountingFrameAllocator {
-    allocator: FrameAllocator,
-    total: usize,
-    allocated: usize,
-}
+/// # PageAlloc trait
+///
+/// The PageAlloc Trait provides the interface for the page allocator.
+/// PageAlloc Trait decouples the page allocator implementation from the
+/// `ostd`, which allows the page allocator to be implemented by the users
+/// wherever they want.
+///
+/// You can refer to `kernel/libs/page_allocator` for the example
+/// implementation.
+pub trait PageAlloc: Sync + Send {
+    /// Add a range of **frame number** [start, end) to the allocator
+    ///
+    /// Warning! May lead to panic when afterwards allocation while using
+    /// out-of `ostd`
+    fn add_frame(&mut self, start: usize, end: usize);
 
-impl CountingFrameAllocator {
-    pub fn new(allocator: FrameAllocator, total: usize) -> Self {
-        CountingFrameAllocator {
-            allocator,
-            total,
-            allocated: 0,
-        }
-    }
+    /// Allocates a contiguous range of pages described by the layout.
+    ///
+    /// # Panics
+    ///
+    /// The function panics if the layout.size is not base-page-aligned or
+    /// if the layout.align is less than the PAGE_SIZE.
+    fn alloc(&mut self, layout: Layout) -> Option<Paddr>;
```

Comment:
We've just decided that the page allocator trait should enable lock-free design of local page allocation cache. So the signature should be `fn alloc(&self, layout: Layout) -> Option<Paddr>;`. Similarly, the `alloc_page` and `dealloc` functions takes shared references. The implementation should manage it's own locking scheme. For now, you can make a inner big lock around the global free list, and make the statistics atomic counters.

In the future, we can maintain per-cpu free lists to allow lock-free design.

Reply:
Get it. I will refactor the corresponding trait in the subsequent pr. Next pr will furtherly enhance the performance of page allocator, as well as refactor the corresponding trait to meet the latest design.